<a href="https://colab.research.google.com/github/JayShekhavat/Twitter-Sentiment-Analysis-Project-NLP-/blob/main/Twitter_Sentiment_Analysis_Project_(NLP).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Twitter Sentiment Analysis (NLP) Project

##Import some common libraries

In [1]:
import numpy as np
import pandas as pd


##Import data in Excel format

In [2]:
df = pd.read_excel("/content/dataset.xlsx")

##show 5 rows of dataset

In [3]:
df.head()

text  label
0                                         oh my gosh    1.0
1  trouble sleeping, confused mind, restless hear...    1.0
2  All wrong, back off dear, forward doubt. Stay ...    1.0
3  I've shifted my focus to something else but I'...    1.0
4  I'm restless and restless, it's been a month n...    1.0

##Shape of dataset

In [4]:
df.shape

(6982, 2)

There are 6982 rows and 2 columns.

##Unique values of label column.

In [5]:
df['label'].value_counts()

0.0    6247
1.0     733
Name: label, dtype: int64

There are 6247 records if 0's and 733 records of 1's.

##Information of dataset

In [6]:
df['label'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 6982 entries, 0 to 6981
Series name: label
Non-Null Count  Dtype  
--------------  -----  
6980 non-null   float64
dtypes: float64(1)
memory usage: 54.7 KB


##Check null values in dataset

In [7]:
df.isnull().sum()

text     10
label     2
dtype: int64

There are 10 null values in text column and 2 null values in label column.

##Drop null values from dataset

In [8]:
df.dropna( inplace  = True)

In [9]:
df.isnull().sum()

text     0
label    0
dtype: int64

In [10]:
df.shape

(6970, 2)

##Check duplicates values in dataset

In [11]:
df.duplicated().sum()

77

There are 77 duplicates values in dataset

In [12]:
df.drop_duplicates(inplace  = True)

In [13]:
df.duplicated().sum()

0

##Install nltk library

In [14]:
!pip install nltk

##Import nltk library

In [15]:
import nltk

##Import all needed libraries for NLP Project

In [16]:
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


##Download stopwords

In [17]:
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

##Length of stopwords and print them

In [18]:
print(len(stopwords.words('english')))

179


In [19]:
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

##Text and label columns assigne to text variable and label variable respectivily

In [20]:
text = df['text']

In [21]:
target = df['label']

In [22]:
text

0                                              oh my gosh
1       trouble sleeping, confused mind, restless hear...
2       All wrong, back off dear, forward doubt. Stay ...
3       I've shifted my focus to something else but I'...
4       I'm restless and restless, it's been a month n...
                              ...                        
6977                    I can't forget you #SpiritHadrian
6978    € ®šæœŸâ˜†ã€'..DJ DAIKI! DJ DAIKI! DJ DAIKI!.D...
6979                                            Dai5y! <3
6980    tired of clowns but still hopefully tonight if...
6981                                   MW SUBI WN LA VACA
Name: text, Length: 6893, dtype: object

In [23]:
text.info()

<class 'pandas.core.series.Series'>
Int64Index: 6893 entries, 0 to 6981
Series name: text
Non-Null Count  Dtype 
--------------  ----- 
6893 non-null   object
dtypes: object(1)
memory usage: 107.7+ KB


In [24]:
target

0       1.0
1       1.0
2       1.0
3       1.0
4       1.0
       ... 
6977    0.0
6978    0.0
6979    0.0
6980    0.0
6981    0.0
Name: label, Length: 6893, dtype: float64

## Stage 1 ->data convert to lower case

In [25]:
text = text.tolist()

In [26]:
len(text)

6893

In [27]:
lower_case_stage_1 = []
def lower_case(text):
  for sent in text:
    lower_case_stage_1.append(str.lower(sent))


In [28]:
print(type(lower_case_stage_1))

<class 'list'>


In [29]:
len(lower_case_stage_1)

0

In [30]:
lower_case(text)

In [31]:
lower_case_stage_1

['oh my gosh',
 'trouble sleeping, confused mind, restless heart. all out of tune',
 'all wrong, back off dear, forward doubt. stay in a restless and restless place',
 "i've shifted my focus to something else but i'm still worried",
 "i'm restless and restless, it's been a month now, boy. what do you mean?",
 'every break, you must be nervous, like something is wrong, but what the heck',
 'i feel scared, anxious, what can i do? and may my family or us be protected :)',
 "have you ever felt nervous but didn't know why?",
 "i haven't slept well for 2 days, it's like i'm restless. why huh :([].",
 "i'm really worried, i want to cry.",
 "always restless every night, even though i don't know why, what's wrong. strange.",
 "i'm confused, i'm not feeling good lately. every time i want to sleep, i always feel restless",
 'sometimes what is needed when there is a problem is to laugh until you forget that there is a problem, when you remember it, you feel restless like that well, it turns out th

##stage 2 -> Tokenization

In [32]:
from nltk.tokenize import sent_tokenize,word_tokenize
nltk.download('punkt')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [33]:
token_stage_2 = []
for sent in lower_case_stage_1:
    sent = word_tokenize(sent)
    token_stage_2.append(sent)

In [34]:
token_stage_2

[['oh', 'my', 'gosh'],
 ['trouble',
  'sleeping',
  ',',
  'confused',
  'mind',
  ',',
  'restless',
  'heart',
  '.',
  'all',
  'out',
  'of',
  'tune'],
 ['all',
  'wrong',
  ',',
  'back',
  'off',
  'dear',
  ',',
  'forward',
  'doubt',
  '.',
  'stay',
  'in',
  'a',
  'restless',
  'and',
  'restless',
  'place'],
 ['i',
  "'ve",
  'shifted',
  'my',
  'focus',
  'to',
  'something',
  'else',
  'but',
  'i',
  "'m",
  'still',
  'worried'],
 ['i',
  "'m",
  'restless',
  'and',
  'restless',
  ',',
  'it',
  "'s",
  'been',
  'a',
  'month',
  'now',
  ',',
  'boy',
  '.',
  'what',
  'do',
  'you',
  'mean',
  '?'],
 ['every',
  'break',
  ',',
  'you',
  'must',
  'be',
  'nervous',
  ',',
  'like',
  'something',
  'is',
  'wrong',
  ',',
  'but',
  'what',
  'the',
  'heck'],
 ['i',
  'feel',
  'scared',
  ',',
  'anxious',
  ',',
  'what',
  'can',
  'i',
  'do',
  '?',
  'and',
  'may',
  'my',
  'family',
  'or',
  'us',
  'be',
  'protected',
  ':',
  ')'],
 ['have',


##Remove Special characters from dataset

In [35]:
len(token_stage_2)

6893

In [36]:
import re

In [37]:
removed_spl_char = []
for words in token_stage_2:
  clean = []
  for word in words:
    res = re.sub(r'[^\w\s]', "",word)
    if res !="":
      clean.append(res)
  removed_spl_char.append(clean)



In [38]:
removed_spl_char

[['oh', 'my', 'gosh'],
 ['trouble',
  'sleeping',
  'confused',
  'mind',
  'restless',
  'heart',
  'all',
  'out',
  'of',
  'tune'],
 ['all',
  'wrong',
  'back',
  'off',
  'dear',
  'forward',
  'doubt',
  'stay',
  'in',
  'a',
  'restless',
  'and',
  'restless',
  'place'],
 ['i',
  've',
  'shifted',
  'my',
  'focus',
  'to',
  'something',
  'else',
  'but',
  'i',
  'm',
  'still',
  'worried'],
 ['i',
  'm',
  'restless',
  'and',
  'restless',
  'it',
  's',
  'been',
  'a',
  'month',
  'now',
  'boy',
  'what',
  'do',
  'you',
  'mean'],
 ['every',
  'break',
  'you',
  'must',
  'be',
  'nervous',
  'like',
  'something',
  'is',
  'wrong',
  'but',
  'what',
  'the',
  'heck'],
 ['i',
  'feel',
  'scared',
  'anxious',
  'what',
  'can',
  'i',
  'do',
  'and',
  'may',
  'my',
  'family',
  'or',
  'us',
  'be',
  'protected'],
 ['have', 'you', 'ever', 'felt', 'nervous', 'but', 'did', 'nt', 'know', 'why'],
 ['i',
  'have',
  'nt',
  'slept',
  'well',
  'for',
  '2'

##Remove Stopwords from dataset

In [39]:
removed_stop_words = []
for words in removed_spl_char:
  w = []
  for  word in words:
    if not word in stopwords.words('english'):
      w.append(word)
  removed_stop_words.append(w)




In [40]:
removed_stop_words

[['oh', 'gosh'],
 ['trouble', 'sleeping', 'confused', 'mind', 'restless', 'heart', 'tune'],
 ['wrong',
  'back',
  'dear',
  'forward',
  'doubt',
  'stay',
  'restless',
  'restless',
  'place'],
 ['shifted', 'focus', 'something', 'else', 'still', 'worried'],
 ['restless', 'restless', 'month', 'boy', 'mean'],
 ['every', 'break', 'must', 'nervous', 'like', 'something', 'wrong', 'heck'],
 ['feel', 'scared', 'anxious', 'may', 'family', 'us', 'protected'],
 ['ever', 'felt', 'nervous', 'nt', 'know'],
 ['nt', 'slept', 'well', '2', 'days', 'like', 'restless', 'huh'],
 ['really', 'worried', 'want', 'cry'],
 ['always',
  'restless',
  'every',
  'night',
  'even',
  'though',
  'nt',
  'know',
  'wrong',
  'strange'],
 ['confused',
  'feeling',
  'good',
  'lately',
  'every',
  'time',
  'want',
  'sleep',
  'always',
  'feel',
  'restless'],
 ['sometimes',
  'needed',
  'problem',
  'laugh',
  'forget',
  'problem',
  'remember',
  'feel',
  'restless',
  'like',
  'well',
  'turns',
  'stil

##Stemming the words of sentences of dataset.

In [41]:
stemming = PorterStemmer()

In [42]:
stemmed_text = []
for sent in removed_stop_words:
  w = []
  for word in sent:
    w.append(stemming.stem(word))
  stemmed_text.append(w)


In [43]:
stemmed_text

[['oh', 'gosh'],
 ['troubl', 'sleep', 'confus', 'mind', 'restless', 'heart', 'tune'],
 ['wrong',
  'back',
  'dear',
  'forward',
  'doubt',
  'stay',
  'restless',
  'restless',
  'place'],
 ['shift', 'focu', 'someth', 'els', 'still', 'worri'],
 ['restless', 'restless', 'month', 'boy', 'mean'],
 ['everi', 'break', 'must', 'nervou', 'like', 'someth', 'wrong', 'heck'],
 ['feel', 'scare', 'anxiou', 'may', 'famili', 'us', 'protect'],
 ['ever', 'felt', 'nervou', 'nt', 'know'],
 ['nt', 'slept', 'well', '2', 'day', 'like', 'restless', 'huh'],
 ['realli', 'worri', 'want', 'cri'],
 ['alway',
  'restless',
  'everi',
  'night',
  'even',
  'though',
  'nt',
  'know',
  'wrong',
  'strang'],
 ['confus',
  'feel',
  'good',
  'late',
  'everi',
  'time',
  'want',
  'sleep',
  'alway',
  'feel',
  'restless'],
 ['sometim',
  'need',
  'problem',
  'laugh',
  'forget',
  'problem',
  'rememb',
  'feel',
  'restless',
  'like',
  'well',
  'turn',
  'still',
  'sad',
  'burden',
  'blah',
  'blah',

##Lemmetized the words of sentences of dataset.

In [44]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [45]:
from nltk.stem.wordnet import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()


In [46]:
lemmetized_text = []
for sent in stemmed_text:
  w = []
  for word in sent:
    w.append(lemmatizer.lemmatize(word, pos = 'a'))
  lemmetized_text.append(" ".join(w))

In [47]:
df['text'] = lemmetized_text

In [48]:
df['text'].value_counts()

                                                                 18
nervou                                                           13
morn                                                             10
sleepi                                                            9
say dreami                                                        7
                                                                 ..
watch coverag explos beirut last year morn realli sad             1
peak human satur long think ration life                           1
type messag nt yet sent                                           1
wrong checkout bag ask accept cancel order seller nt give itâ     1
mw subi wn la vaca                                                1
Name: text, Length: 6749, dtype: int64

In [49]:
df['text'].isnull().sum()

0

In [50]:
x = df['text']

In [51]:
x.head()

0                                              oh gosh
1         troubl sleep confus mind restless heart tune
2    wrong back dear forward doubt stay restless re...
3                    shift focu someth els still worri
4                     restless restless month boy mean
Name: text, dtype: object

In [52]:
target.head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: label, dtype: float64

In [53]:
y = target

In [54]:
y.head()

0    1.0
1    1.0
2    1.0
3    1.0
4    1.0
Name: label, dtype: float64

##Split the data into train test for training and test.

In [55]:

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=42)

In [56]:
x_train.shape

(5169,)

In [57]:
x_train.head()

6901    look mutual esp nctzen wayzenni fandom sabi he...
5741                   kuy wtb tf 213 fund bni drop clean
968       ca nt jealou choos someth els right need expect
898                                                  woke
3401                                               fhcukh
Name: text, dtype: object

In [58]:
x_test.head()

1174    realli understand feel struggl product problem...
6246                           end day grate oh god sorri
1607                       principl dp repay think keriðÿ
2792    sooo cold runni nose sneez wo nt stop pleas pl...
5607    hlt hemorrhoid hemorrhoid exampl go doctor sor...
Name: text, dtype: object

In [59]:
y_train.head()

6901    0.0
5741    0.0
968     0.0
898     0.0
3401    0.0
Name: label, dtype: float64

In [60]:
y_test.head()

1174    0.0
6246    0.0
1607    0.0
2792    0.0
5607    0.0
Name: label, dtype: float64

##Vectorize the dataset

In [61]:
vectorizer = TfidfVectorizer()
x_train = vectorizer.fit_transform(x_train)
x_test = vectorizer.transform(x_test)

In [64]:
print(x_train)

  (0, 7312)	0.3588110015126823
  (0, 6266)	0.30039243186654235
  (0, 3286)	0.2526247144018736
  (0, 6305)	0.3652239766400849
  (0, 2591)	0.3286214586755081
  (0, 7853)	0.343516951051947
  (0, 5070)	0.3198652884152557
  (0, 2462)	0.3726272131088615
  (0, 4969)	0.2612494865677704
  (0, 4435)	0.21999808398057802
  (1, 1648)	0.28442753235071055
  (1, 2284)	0.28442753235071055
  (1, 1206)	0.40968206208796976
  (1, 2822)	0.34556225802790747
  (1, 187)	0.40968206208796976
  (1, 7261)	0.3673011768751352
  (1, 8041)	0.34556225802790747
  (1, 4219)	0.3588683195935763
  (2, 2522)	0.37513601601707125
  (2, 5085)	0.31046371433046044
  (2, 6195)	0.3000041092296803
  (2, 2394)	0.38299664400887706
  (2, 6800)	0.3512673936964928
  (2, 1597)	0.3776413841544467
  (2, 3808)	0.402803844620402
  :	:
  (5165, 8192)	0.25215905707701625
  (5165, 1080)	0.25215905707701625
  (5165, 4307)	0.6380800714214172
  (5165, 288)	0.2406160798132205
  (5165, 4443)	0.2208832298449486
  (5165, 7745)	0.23242620710874434
  (51

In [65]:
print(x_test)

  (0, 8123)	0.31286456936107837
  (0, 7612)	0.1681619032656141
  (0, 6977)	0.20967717494343402
  (0, 6056)	0.11236527781017186
  (0, 5826)	0.4041745837811421
  (0, 5817)	0.1742748215481239
  (0, 5800)	0.19432673696255537
  (0, 5285)	0.38865347392511074
  (0, 4797)	0.2127251281159261
  (0, 4529)	0.1287551856610564
  (0, 4445)	0.14952142870013962
  (0, 2964)	0.1831699846007921
  (0, 2940)	0.12215913771087812
  (0, 2681)	0.14579444845518416
  (0, 2636)	0.11872062587243082
  (0, 2413)	0.1674811942373417
  (0, 2106)	0.20440283128656955
  (0, 1723)	0.21999535267955195
  (0, 1707)	0.12374665213399132
  (0, 1700)	0.19608053445721302
  (0, 1577)	0.1868680505015826
  (0, 254)	0.24450666211543245
  (1, 6822)	0.4286254355573382
  (1, 5289)	0.3703654692947165
  (1, 3047)	0.4967266705261268
  :	:
  (1718, 7344)	0.27255158641807076
  (1718, 7292)	0.2714056390915975
  (1718, 6801)	0.326326008068181
  (1718, 6163)	0.34917885756672407
  (1718, 5223)	0.1572373350611522
  (1718, 2486)	0.45969287811053794


##Applied Logistic Regression in dataset  

In [66]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
model = log_reg.fit(x_train, y_train)



##Predict the output

In [67]:
y_predict =model.predict(x_test)

##Show accuracy

In [68]:
accuracy = accuracy_score(y_predict, y_test)

In [69]:
accuracy

0.9675174013921114

Accuracy is 0.9675174013921114 Which is very good accuracy.